docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+- Using MySQL compatible system variables (`SET GLOBAL sysvar=x`, `SET SESSION sysvar=x`) ([Docs](https://docs.pingcap.com/tidb/stable/system-variables))
+
+Both the **semantics** and **naming conventions** differ between these two methods:
+
+1. Configuration file settings only apply to a single TiDB server instance. Making changes to settings managed by the configuration file requires restarting the TiDB server.
+2. System variables _natively_ manage settings that have _session_ (connection) or _global_ (cluster-wide) scope. However, some _session_ scoped variables are used to allow instance configuration. This is not a feature that is natively supported, and the usage is often confusing.
+3. The configuration file uses a hierarchy of sections such as `[performance]` or `[experimental]`.
+4. System variables use a flat naming convention (but often use a prefix of `tidb_feature_XXX`). Thus mapping between the two is not straight forward.
+
+This proposal introduces _native_ support for `INSTANCE` scoped variables, such that system variables offer the superset of functionality of configuration files. It does this using a much simplified implementation from earlier proposals: an individual system variable **must not** permit both `GLOBAL` and `INSTANCE` scope. Implementing this restriction is key to this proposal; since it avoids a confusing set of precedence rules which are hard to understand. For alternative proposals see "investigation and alternatives".
+
+## Motivation or Background
+
+The motivation for this proposal is ease of use and maintainability:
+
+1. It's not clear if each setting should be a configuration file setting or a system variable (we usually choose system variables, but there is no clear rule). After this proposal is implemented, we can make every setting available as a system variable (even if read-only).
+2. We are being asked to make additional configuration file settings dynamically configuration (via a system variable). When we do this, the naming is inconsistent. After this proposal is implemented, we can allow the `.toml` file to support an `[instance]` section where the flat named for system variables can be used. This makes it identical to how system variables are configured in MySQL.
+3. The current system is hard for users. It's hard to explain the two systems, and hard to explain the differences between them. Often "INSTANCE" scoped system variables are incorrectly documented as `SESSION`, since it's not a native behavior. Explaining how `SESSION` scope is hijacked is difficult for MySQL users to understand, because usually changes in a different session should not affect your session.
+
+## Detailed Design
+
+### Stage 1: Initial Implementation
+
+From a user-oriented point of view, setting an instance scoped sysvar is the same as setting a `GLOBAL` variable:
+
+```sql
+SET GLOBAL max_connections=1234;
```

Comment:
Will this confuse the customer that `SET GLOBAL` is used by instance and global sysvars? How could they know if a var is instance or global level?

Reply:
Yes, you will need to read the docs to see if a variable is `INSTANCE` or `GLOBAL` scoped. This confusion is existing with the session scope, but actually worse since it's not handled by the framework the docs for many `INSTANCE` variables say they are `SESSION` scoped (we will finally be able to auto generate docs and say something is "INSTANCE" scoped versus a manual workaround: https://github.com/pingcap/docs/blob/8600643f2fcd20d70d8ad8f637009995948c48c6/scripts/generate-system-variables.go#L142-L152 ).

The semantics for `INSTANCE` are also much closer to global than session, with only two exceptions:
- Changes don't propagate to other tidb servers.
- Changes don't persist.

Reply:
I have added a section on "Documentation Changes", PTAL